proposals/gc/MVP.md

```diff
@@ -266,62 +293,112 @@ Tentatively, support a type of guaranteed unboxed scalars.
+  - traps if reference is not compound data
+  - equivalent to `(block $l (param anyref) (result dataref) (br_on_data $l) (unreachable))`
 
 * `ref.as_i31` converts to an integer reference
   - `ref.as_i31 : [anyref] -> [i31ref]`
   - traps if reference is not an integer
-  - equivalent to `(rtt.canon i31) (ref.cast)`
+  - equivalent to `(block $l (param anyref) (result i31ref) (br_on_i31 $l) (unreachable))`
+
+Note: There are no instructions to check for `externref`, since that can consist of a diverse set of different object representations that would be costly to check for exhaustively.
+
+TODO: Should we add `br_on_null` for completeness?
 
 
 #### Runtime Types
 
-* `rtt.canon <heaptype>` returns the RTT of the specified type
-  - `rtt.canon t : [] -> [(rtt n t)]`
-  - `n = 0` iff `t = any`, and `n = 1` otherwise
+* `rtt.canon <typeidx>` returns the RTT of the specified type
+  - `rtt.canon $t : [] -> [(rtt 0 $t)]`
   - multiple invocations of this instruction yield the same observable RTTs
   - this is a *constant instruction*
-  - equivalent to `(rtt.sub 1 any t (rtt.canon any))`, except when `t` itself is `any`
 
-* `rtt.sub <n> <heaptype1> <heaptype2>` returns an RTT for `heaptype2` as a sub-RTT of a the parent RTT operand for `heaptype1`
-  - `rtt.sub n t1 t2 : [(rtt n t1)] -> [(rtt (n+1) t2)]`
+* `rtt.sub <n> <typeidx1> <typeidx2>` returns an RTT for `typeidx2` as a sub-RTT of a the parent RTT operand for `typeidx1`
+  - `rtt.sub n $t1 $t2 : [(rtt n $t1)] -> [(rtt (n+1) $t2)]`
     - iff `t2 <: t1`
   - multiple invocations of this instruction with the same operand yield the same observable RTTs
   - this is a *constant instruction*
 
-TODO: Add the ability to generate new (non-canonical) RTT values to implement casting in nominal type hierarchies.
+TODO: Add the ability to generate new (non-canonical) RTT values to implement casting in nominal type hierarchies?
 
 
 #### Casts
 
-* `ref.test <heaptype1> <heaptype2>` tests whether a reference value's [runtime type](#values) is a [runtime subtype](#runtime) of a given RTT
-  - `ref.test t1 t2 : [(ref null t1) (rtt n t2)] -> [i32]`
-     - iff `t2 <: t1`
+RTT-based casts can only be performed with respect to concrete types, and require a data or function reference as input, which are known to carry an RTT.
+
+* `ref.test <typeidx>` tests whether a reference value's [runtime type](#values) is a [runtime subtype](#runtime) of a given RTT
+  - `ref.test $t : [(ref null ht) (rtt n $t)] -> [i32]`
+    - iff `ht <: data` or `ht <: func`
+    - and `(type $t) <: ht`
   - returns 1 if the first operand is not null and its runtime type is a sub-RTT of the RTT operand, 0 otherwise
 
-* `ref.cast <heaptype1> <heaptype2>` casts a reference value down to a type given by a RTT representation
-  - `ref.cast t1 t2 : [(ref null t1) (rtt n t2)] -> [(ref t2)]`
-     - iff `t2 <: t1`
+* `ref.cast <typeidx>` casts a reference value down to a type given by a RTT representation
```

Comment:
What about adding the (optional) depth of the RTT as an immediate here? Since there are now two kinds of RTT checks, one where the depth is known and one where the depths are unknown, which kind of check it is depends on the type of the incoming RTT value. An interpreter doesn't know that statically, so it needs to inspect the RTT value to know whether it is a "depth carrying" RTT or a "depth independent" RTT.

Reply:
Yeah, this is part of a wider discussion. In the light of recent discussions, it seems like we rather might want to remove the immediate altogether in cases like this. So I suggest leaving as is until we have resolved that discussion.

For an interpreter, I would assume that it simply ignores the static depth altogether and just reads it off the RTT value uniformly. The win of using the static value may already be tiny for a compiler, I doubt that it matters in an interpreter.

Reply:
I think you would get different results because isn't an RTT with depth a subtype of RTT without? And thus the static type and dynamic type of the RTT may be different, so an interpreter could not know the difference by inspecting the value alone.

Reply:
Yes, `(rtt n $t) <: (rtt $t)`, see the definition of subtyping above. This is so that you can abstract from the depth where you don't want to expose or depend on it. It can't result in a different value, though. It's just the length of the supertype vector in the value, so presumably needs to be derivable from the value representation anyway.

Reply:
I agree. What I mean is that a cast that uses an RTT of static type `(rtt n $t)` checks only that the RTT at index `n` in the supertype vector, whereas a cast with RTT of static type `(rtt $t)` checks all entries in the supertype vector. Because of subtyping on RTT values, an RTT value allocated as `(rtt n $t)` is a subtype of `(rtt $t)` and can end up at cast site of static type `(rtt $t)`, so it is not possible to distinguish the two cases based on the RTT value's dynamic type alone.

Reply:
> whereas a cast with RTT of static type (rtt $t) checks all entries in the supertype vector.

Hm, that's not the intention. Why should it have to do that? The only difference should be how it determines the slot index, everything else should be the same.

Reply:
The difference is that using the types (of RTTs) inferred from validation, a JIT would use the slot index inferred from the RTT's static type and emit a check _only_ of that slot index, and not a search, whereas an interpreter would not have this information from the inferred RTT type, as it only has the RTT values.

Thus I think it's necessary to have the optional depth here, otherwise you force JITs to emit a search too.